### ISSUE TITLE
Questions: BatchNorm and Recurrent Dropout

### ISSUE THREAD

**Issue by sallamander** [Smerity/keras_snli#2]
This is awesome! 💯 

I know you said you didn't any real grid-searching, but out of curiosity:
1. Did you experiment w/o the `BatchNormalization`? I have yet to have it actually help in an RNN, so I'd be curious to know if you tried it without and if that yielded worse results. 
2. Any particular reason why you used dropout only with the input gates (`dropout_W`) when using the rnn models? I've always had more success when combining it with dropout on the recurrent units as well (`dropout_U`), but am curious if you tried the combination and found that it didn't help.

**Comment by Smerity** [comment]
Hey @sallamander :)
1. I did one brief experiment without `BatchNormalization` and the results appeared similar but did not evaluate it properly. I was primarily following the codebase from Bowman which used `BatchNormalization` on almost all the outputs. One interesting note is that you don't want to use a shared `BatchNormalization` layer for both premise and hypothesis given the distribution of those inputs are different (i.e. each premise is repeated many times each).
2. Gah! That was a total oversight. I intended to apply the Keras RNN recurrent dropout (`dropout_U`) which follows [Gal 2015](http://arxiv.org/abs/1512.05287) but just was silly and forgot it. I'm going to run models with recurrent dropout applied and see what impact that has. Hopefully it gets the GRU and/or LSTM up past the summation of word vectors baseline.

**Comment by sallamander** [comment]
1. Ah, okay gotcha, makes sense. 
2. I'll be interested to see how it turns out! With the `train` and `validation` higher for the GRU and LSTM, I'd think that the right amount of regularization would move the `test` above the summation baseline.  At the same time it's still pretty sweet that the summation baseline does that well to begin with.

**Comment by Smerity** [comment]
Single run result shows GRU finally beating `sum(word vectors)` :)

The GRU hit a test accuracy, 0.8321, which is higher than the best `sum(word vectors)` I've seen (@bradleypallen reported 0.8313, and I've seen similar). Uncertain whether it's statistically significant. LSTM didn't do as well (0.8209).

When I do another run or two, I'll add them to the table.

GRU with recurrent dropout:

> Epoch 27/42
> 549367/549367 [==============================] - 333s - loss: 0.4388 - acc: 0.8397 - val_loss: 0.4245 - val_acc: 0.8403
> Epoch 28/42
> 549367/549367 [==============================] - 333s - loss: 0.4370 - acc: 0.8409 - val_loss: 0.4453 - val_acc: 0.8367
> Epoch 29/42
> 549367/549367 [==============================] - 333s - loss: 0.4350 - acc: 0.8416 - val_loss: 0.4387 - val_acc: 0.8335
> Epoch 30/42
> 549367/549367 [==============================] - 335s - loss: 0.4341 - acc: 0.8416 - val_loss: 0.4381 - val_acc: 0.8364
> Epoch 31/42
> 549367/549367 [==============================] - 334s - loss: 0.4331 - acc: 0.8425 - val_loss: 0.4328 - val_acc: 0.8371
> Epoch 32/42
> 549367/549367 [==============================] - 334s - loss: 0.4321 - acc: 0.8435 - val_loss: 0.4269 - val_acc: 0.8377
> Epoch 33/42
> 549367/549367 [==============================] - 334s - loss: 0.4307 - acc: 0.8442 - val_loss: 0.4307 - val_acc: 0.8396
> Epoch 34/42
> 549367/549367 [==============================] - 334s - loss: 0.4278 - acc: 0.8452 - val_loss: 0.4352 - val_acc: 0.8325
> Epoch 35/42
> 549367/549367 [==============================] - 335s - loss: 0.4273 - acc: 0.8450 - val_loss: 0.4303 - val_acc: 0.8377
> Epoch 36/42
> 549367/549367 [==============================] - 335s - loss: 0.4264 - acc: 0.8461 - val_loss: 0.4333 - val_acc: 0.8392
> 9824/9824 [==============================] - 1s
> Test loss / test accuracy = 0.4422 / 0.8321

LSTM with recurrent dropout:

> Epoch 17/42
> [==============================] - 375s - loss: 0.4550 - acc: 0.8297 - val_loss: 0.4415 - val_acc: 0.8307
> Epoch 18/42
> [==============================] - 375s - loss: 0.4499 - acc: 0.8321 - val_loss: 0.4577 - val_acc: 0.8291
> Epoch 19/42
> [==============================] - 375s - loss: 0.4462 - acc: 0.8347 - val_loss: 0.4416 - val_acc: 0.8305
> Epoch 20/42
> [==============================] - 376s - loss: 0.4422 - acc: 0.8361 - val_loss: 0.4497 - val_acc: 0.8279
> Epoch 21/42
> [==============================] - 375s - loss: 0.4391 - acc: 0.8378 - val_loss: 0.4450 - val_acc: 0.8342
> Epoch 22/42
> [==============================] - 375s - loss: 0.4355 - acc: 0.8395 - val_loss: 0.4487 - val_acc: 0.8330
> Epoch 23/42
> [==============================] - 375s - loss: 0.4322 - acc: 0.8413 - val_loss: 0.4442 - val_acc: 0.8333
> Epoch 24/42
> [==============================] - 375s - loss: 0.4295 - acc: 0.8428 - val_loss: 0.4440 - val_acc: 0.8361
> Epoch 25/42
> [==============================] - 375s - loss: 0.4275 - acc: 0.8435 - val_loss: 0.4440 - val_acc: 0.8349
> Epoch 26/42
> 241152/549367 [============>.................] - ETA: 209s - loss: 0.4159 - acc: 0.8486
> [==============================] - 375s - loss: 0.4237 - acc: 0.8453 - val_loss: 0.4445 - val_acc: 0.8371
> 9824/9824 [==============================] - 2s
> Test loss / test accuracy = 0.4625 / 0.8209

**Comment by Smerity** [comment]
Second run using GRU and recurrent dropout:

> Epoch 42/42
> 549367/549367 [==============================] - 332s - loss: 0.4216 - acc: 0.8483 - val_loss: 0.4352 - val_acc: 0.8395
> 9824/9824 [==============================] - 2s
> Test loss / test accuracy = 0.4455 / 0.8308

The validation loss is better for the first run (0.4245 < 0.4352), so the 83.2 still stands.
Only the BiLSTM encoders are consistently better on the SNLI scoreboard.
I still need to run more LSTM experiments - regularization may need to be stronger due to the larger number of parameters..?

**Comment by sallamander** [comment]
Are the parameters you're using still the same as the script currently shown in the repo.? 

If so, I think stronger regularization might help, particularly a higher dropout rate. I've often found something close to 50% dropout helps, but given how close your `loss` and `val_loss` are here, it looks like something close to the 20% might be optimal given the current set up of your networks.

How many layers have you tried for these (I can't really tell given the script, but sorry if I missed it)? The dataset size is pretty large, right (570k examples overall)? Given that, if the goal is to kind of just play around and see how different models do, I'd be tempted to throw 2-3 encoding/decoding layers each at it, of 512 + nodes. I don't know how that compares with your current setup, though.

On a somewhat different note, I'm kind of pumped to see the `GRU` networks outperforming the `LSTM` networks - is this consistent, and do you see this with other problems you've done working with RNNs? I've only done a handful of RNN projects, but each time `GRU` units have outperformed `LSTM` units.
